### PR TITLE
feat(medcat-plugins) trainable linker for the embedding linker

### PR DESCRIPTION
**Pull request by adam-sutton-1992**
Hihi,

Here's the PR for the trainable embedding linker as foretold by our ancestors.

I'll list the main parts of it here, then the performances.

1. Creates a new module `trainable_embedding_linker` that inherits from the `embedding_linker`. The only functionality in trainable version is that which is only used by it.
2. Split out the modelling functionality from the `embedding_linker` into the `transformer_context_model` module. This has two classes within it: `ContextModel` which handles all embedding tasks (`embed_cuis`...). `ModelForEmbeddingLinking` inherits from the transformers `nn.module` and is where the model logic is. It's largely a wrapper around the language model with a few choices of how to embed.
3. Added additional variables to the `config` these explain themselves, all of these are to do with training and effectively are trade offs between time / compute / performance.

Onto performances:

This is trained / tested with three datasets that all contain SNOMED CT labels. The SNOMED CT Entity Benchmark, Distemist, and COMETA. Due to this requiring training I have a train/test split of 80/20. With the implementation of cat.train at the time of making this I kind of built the training loop outside of the medcat library. 

Effectively it looks like this:

```
cat.trainer.train_supervised_raw(train_projects, test_size=0, nepochs=1)
cat._pipeline._components[-1].train_on_batch()
cat._pipeline._components[-1].refresh_structure()
cat._pipeline._components[-1].create_embeddings()
get_stats(cat=cat, data=test_projects, use_project_filters=False)
for i in range(7):
    cat.trainer.train_supervised_raw(train_projects, test_size=0, nepochs=1)
    cat._pipeline._components[-1].train_on_batch()
    cat._pipeline._components[-1].create_embeddings()
    get_stats(cat=cat, data=test_projects, use_project_filters=False)
```

Here's the baseline performance without training (which would be comparing to a normal embedding_linker):

Epoch: 0, Prec: 0.081820475543968, Rec: 0.3308877119673485, F1: 0.13119870535198244

Here's the best performance with a few hyperparams I've found to be optimal:

Epoch: 0, Prec: 0.11907464089601173, Rec: 0.5048605035481676, F1: 0.19269978201382124

These hyper-params I mention are, with a bit of commentary:

`cat.config.components.linking.embedding_model_name = "abhinand/MedEmbed-small-v0.1"`  (we default to "sentence-transformers/all-MiniLM-L6-v2" because it's kind of a standard, and it's small (6 layers). But MedEmbed in my experience is the best embedding I've found for medical stuff)
`context_window_size` = 35 (with more context performance should increase, if you use the `mention_mask`, the gains past 14 tokens is very marginal)
`cat._pipeline._components[-1].context_model.model.unfreeze_top_n_lm_layers(4)` (As you increase the number of layers the performance will increase, at the cost of storing more gradients and computational complexity. You will also transform the embedding space, requiring more data. I haven't gone beyond unfreezing four layers, because that's a redundant task at this point.)
`lr=1e-4, weight_decay=0.01` (These are hardcoded and it's TODO: it shouldn't be. Oops. If you're only training the linear projection layer you can set it as high as 1e-3. When you start affecting transformer layers it significantly impacts performance for the worse).
`cat._pipeline._components[-1].cnf_l.train_on_names = True` (You can train on cuis or all potential names. A full CDB is about 3 mil names, or 600k CUIs. Names performs slightly better, CUIs performs quite a bit faster. It's a one time cost for training a model, so something to consider`

Here's some earlier iterations performances:

Without mention_attention functionality:
`
Epoch: 0, Prec: 0.10870886017906067, Rec: 0.4608991494532199, F1: 0.17592386464826354
`
It performed best with smaller context windows (size 10).

An earlier experiment: Unfreezing layers (these were done without the COMETA dataset, so just Distemist and the Snomed_CT Benchmark):

ALL LAYERS FROZEN
`
Epoch: 0, Prec: 0.10861838458277627, Rec: 0.40456670917592763, F1: 0.17125743885040035
`
UNFREEZING TOP 1 LAYERS
`
Epoch: 0, Prec: 0.11208035088291403, Rec: 0.4174662941819507, F1: 0.1767163258223325
`
UNFREEZING TOP 2 LAYERS
`
Epoch: 0, Prec: 0.11370157819225252, Rec: 0.4235394145511964, F1: 0.17927559702835402
`
UNFREEZING TOP 4 LAYERS
`
Epoch: 0, Prec: 0.11482145768791782, Rec: 0.4276691364022835, F1: 0.18103758547997326
`
TODO:

1) Test saving and loading trained models perform the same as trained and tested from live.
2) Fix giving LR and weight decay config parameters.
3) Test on another embedding model (i.e. "sentence-transformers/all-MiniLM-L6-v2") - to check performances

### CODE REVIEW

**Review by mart-r — APPROVED** [review]
I must say, some of the details her go over my head.

But I think overall it looks really good!

I did leave a few comments / questions. So feel free to address what's relevant.

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py [hunk 1]
```diff
@@ -499,16 +414,36 @@ def _inference(
...
+                if best_idx < 0 or best_idx >= names_scores.shape[1]:
+                    logger.warning(
+                        "Skipping entity '%s': single-candidate index %s is out of "
+                        "bounds for names_scores width %s.",
+                        entity.detected_name,
+                        best_idx,
+                        names_scores.shape[1],
+                    )
+                    continue
+                similarity = cui_scores[i, best_idx].item()
```
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py:426]: I think I re-edited this about 3/4 times while making the embedding linker and scanning over it.

`best_idx` is sourced from:

```
if len(link_candidates) == 1:
    best_idx = self._cui_to_idx[link_candidates[0]]
    predicted_cui = link_candidates[0]
```

So it should be a cui. 

Essentially here I'm only scoring vs the preferred / longest name in a CUI. Which _could_ be bad if I set a threshold and commonly you would choose the highest scoring name.

I think the confusion I've caused here is that in the case of `elif len(link_candidates) > 1:` then I do consider all potential names and go with the highest scoring.

```
indexed_scores = names_scores[i, name_idxs]
best_local_pos = int(torch.argmax(indexed_scores).item())
best_global_idx = name_idxs[best_local_pos]
similarity = names_scores[i, best_global_idx].item()
best_name = self._name_keys[best_global_idx]
cuis = name_to_cuis[best_name]
if len(cuis) == 1:
    predicted_cui = cuis[0]
else:
    predicted_cui, _ = self._disambiguate_by_cui(cuis, cui_scores[i, :])
```

I think this should be fixed. So I always score and choose the highest scoring names for all length of `link_candidates`.

Not in the PR. I'll fix it immediately after this change :).

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py [hunk 2]
```diff
@@ -654,23 +574,13 @@ def _pre_inference(self, doc: MutableDocument
...
-        self._load_transformers(self.cnf_l.embedding_model_name)
-        if self.cnf_l.train:
+    def predict_entities(
+        self, doc: MutableDocument, ents: list[MutableEntity] | None = None
+    ) -> list[MutableEntity]:
+        if self.cnf_l.train and self.comp_name == "embedding_linker":
             logger.warning(
-                "Attemping to train an embedding linker. This is not required."
+                "Attemping to train a static embedding linker. "
+                "This is not possible / required."
```
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py:583]: Done!

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+            self.train_on_batch()
+            self.training_batch = []
+            self.number_of_batches += 1
+        # If you've got as many batches as you want before re-embedding, 
+        # then do it and reset the counter.
+        if (
+            self.cnf_l.embed_per_n_batches > 0
+            and self.number_of_batches > self.cnf_l.embed_per_n_batches
+        ):
+            print(
```
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: This was me debugging lazily. Sorry!

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/transformer_context_model.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+        """
+        target_model = embedding_model_name or self.cnf_l.embedding_model_name
+        self._refresh_cdb_keys()  # ensure _cui_keys is up to date before embedding
+        self.load_transformers(target_model)
+
+        # TODO: test preffered names instead of longest names
+        # cui_names = [
+        # 	max(self.cdb.cui2info[cui]["names"], key=len) for cui in self._cui_keys
+        # ]
+        cui_names = [self.cdb.get_name(cui) for cui in self._cui_keys]
```
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/transformer_context_model.py:330]: Yes. The TODO that is there currently was actually tested. get_name for these cui embeddings performs better under any training circumstances. 

In the context of the static embedding linker the longest name _slightly_ outperformed `get_name`. When training comes into action it seems that `get_name` is a stronger option. I decided not to make this a configurable variable, due to the impact being so small.

File: medcat-plugins/embedding-linker/tests/__init__.py [hunk 5]
```diff
@@ -6,19 +6,19 @@
 
 RESOURCES_PATH = os.path.join(os.path.dirname(__file__), "resources")
 EXAMPLE_MODEL_PACK_ZIP = os.path.join(RESOURCES_PATH, "mct2_model_pack.zip")
-UNPACKED_EXAMPLE_MODEL_PACK_PATH = os.path.join(
-    RESOURCES_PATH, "mct2_model_pack")
+UNPACKED_EXAMPLE_MODEL_PACK_PATH = os.path.join(RESOURCES_PATH, "mct2_model_pack")
```
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/tests/__init__.py]: I think this was a mypy/linting fix that went into that file by mistake. I can revert it if you want to.

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py [hunk 6]
```diff
@@ -0,0 +1,397 @@
...
+        cdb: CDB,
+        vocab: Vocab,
+        model_load_path: Optional[str],
+    ) -> "TrainableEmbeddingLinker":
+        return cls(cdb, cdb.config)
+
+    def serialise_to(self, folder_path: str) -> None:
+        # Ensure final partial batch is not dropped before saving model state.
+        logger.info("Flushing final training batch before saving model.")
+        logger.info("This is grandfathered in from trainer.py restraints.")
```
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: I think I have a strong preference.

I would like to remove all calls to "train_on_batch" where possible.

This would remove having to call `cat._pipeline._components[-1].train_on_batch()` in the training loop as such in my original PR comment:

```
cat.trainer.train_supervised_raw(train_projects, test_size=0, nepochs=1)
cat._pipeline._components[-1].train_on_batch()
cat._pipeline._components[-1].refresh_structure()
cat._pipeline._components[-1].create_embeddings()
```

I agree for cases of small documents (i.e. COMETA), it's a minor pain to be embedding / training on just a few entities. But I don't think most training will happen this way. 

Preferably I would change the dependency of the embedding linker to MedCAT 2.7 (or whichever one houses these fixes), remove all of these calls, and then protect `train_on_batch` to `_train_on_batch` as originally intended.

If possible I'd like to have it so users don't have to bother with calling `refresh_structure` or `create_embeddings` either after training. But that's probably for a later date.

[the remark this answers is not in the document]
Comment (mart-r) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: Yeah, 2.7 should have this fix in. So I'm fine with supporting only that and above. I think it's unlikely people will be pairing this with old medcat since there isn't really a massive legacy userbase with old versions installed anyway.

File: medcat-plugins/embedding-linker/tests/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mart-r) [reply medcat-plugins/embedding-linker/tests/__init__.py]: Normally we don't lint or check types in test files. But perhaps you ran a script to fix all rather than fix only stuff in the source dir.

I'd prefer this to be rolled back. But I'm fine if it's kept in as well. Not a huge issue.

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mart-r) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py:426]: My main gripe on this was that the `best_idx` comes from CUI based stuff. And then in the check the length of `names_scores` is checked instead. And then it's back to `cui_scores` in the end. So it just feels like the `if` (along with the logged message) should also use `cui_scores`?

[the remark this answers is not in the document]
Comment (mart-r) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py:426]: Ultimately, it's possible it's the way it's supposed to be. But it's just confusing when reading it.
If you do think it's correct, please add a comment on why this is the way that it is if at all possible.

[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/embedding_linker.py:426]: Yeah, sorry. MIssed that part of the logger warning and check. I've corrected it now to cui_scores.

File: medcat-plugins/embedding-linker/tests/__init__.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/tests/__init__.py]: I've reverted.

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: I've done these changes. And uploaded them. Probably worth a quick check!

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py [hunk 7]
```diff
@@ -326,9 +326,15 @@ def train(
             if positive_cui_idx is None:
                 return
             self.training_batch.append((doc, entity, positive_cui_idx))
-        # Train as soon as the batch is full.
-        if len(self.training_batch) >= self.cnf_l.training_batch_size:
-            self.train_on_batch()
+        if (
+            len(self.training_batch) >= self.cnf_l.training_batch_size 
+            or entity is doc.ner_ents[-1]
```
Comment (mart-r) [inline medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: Technically should be checking `.linked_ents` since this is training for linker.

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py [hunk 8]
```diff
@@ -326,9 +326,15 @@ def train(
...
+        ):
+            logger.debug(
+                "End of document reached; training on final batch of size %s",
```
Comment (mart-r) [inline medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: Technically fires on end of batch within document as well. So the message isn't quite accurate.

**Review by mart-r — APPROVED** [review]
2 minor issues, but I think this looks good overall.

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py (hunk 7 above)
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: Done!

File: medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py (hunk 8 above)
Comment (adam-sutton-1992) [reply medcat-plugins/embedding-linker/src/medcat_embedding_linker/trainable_embedding_linker.py]: Done!

**Comment by adam-sutton-1992** [comment]
Hi before merging,

Just for posterity here's some final results from two different base models with the additional fixes we had previously with documents being being a bit weird from the training data set:

```
MiniLM Context Window 20, No Layers Unfrozen, lr-1e-4

Epoch: 0, Prec: 0.13875564581072092, Rec: 0.5778090979782271, F1: 0.22377398620352162

MiniLM Context Window 30, 4 Layers Unfrozen, lr-1e-4

Epoch: 0, Prec: 0.1492378796013165, Rec: 0.6215125887041898, F1: 0.24068287908447522

MedEmbed Context Window 30, 4 Layers Unfrozen, lr-1e-4

Epoch: 0, Prec: 0.15520259546260853, Rec: 0.6464926352632346, F1: 0.25031291467075734

MedEmbed Context Window 30, No Layers Unfrozen, lr-1e-4

Epoch: 0, Prec: 0.14368070953436807, Rec: 0.5984542847421378, F1: 0.23172694164478697
```

So big wins. Fine tuning with layers unfrozen is a big performance winner. However the pre training is still important.

**Comment by adam-sutton-1992** [comment]
Added documentation, along with the removal of some info lines that were out of date. Let me know if this is sufficient.
